### PR TITLE
Add support for upstream gateways or proxies in front of Azkaban web server

### PR DESCRIPTION
**Pull request by mradamlacey**
The following is a pull request to add support in Azkaban's web server for upstream proxies or gateways that mangle the 'real' client IP address.

The scenario where:

client > gateway > Azkaban web server

Can cause issues if the gateway is load balanced (and therefore has different IP addresses) or uses a connection pool to the Azkaban web server as this will cause subsequent requests to have a different Client IP/port.

The pull request uses the `X-Forwarded-For` header if provided, and strips off the port from the client IP address used for session validation.

### CODE REVIEW

File: azkaban-web-server/src/restli/java/azkaban/restli/ResourceUtils.java [hunk 1]
```diff
@@ -56,4 +59,30 @@ public static User getUserFromSessionId(String sessionId, String ip)
...
+    // This will support scenarios where load balancers or gateways
+    // front the Azkaban web server and a changing Ip address invalidates
+    // the session
+    Map<String, String> headers = context.getRequestHeaders();
+
+    String clientIp = headers.getOrDefault("X-Forwarded-For", null);
+    if(clientIp == null){
+      clientIp = (String) context.getRawRequestContext().getLocalAttr("REMOTE_ADDR");
+    }
+    else{
```
Comment (HappyRay) [inline azkaban-web-server/src/restli/java/azkaban/restli/ResourceUtils.java]: Could you think of a way to reduce the code logic duplication in this method and the other one?

**Comment by mradamlacey** [comment]
How do you execute unit tests for the project?

Not that familiar with gradle (sorry).  

`./gradlew test`

Does that do the trick?

Also, for unit tests I'm thinking of just testing a utility class that implements the logic to parse out the client ip...  I don't think it's worth the effort to spin up the test infrastructure for a servlet which would be more of an integration test.  Let me know if you suggest another route.

**Comment by suvodeep-pyne** [comment]
yes. You can execute tests with `./gradew test`. `./gradlew build` will run tests anyway and build the entire project.

Unit tests are good as long as they capture your changes. We expect the tests to implicitly document the expectation from your change. Of course additional context helps people understand the code / logic better.

**Comment by mradamlacey** [comment]
Updated the pull request with the following:

- Unit tests for ResourceContextUtils class (low level helper used to parse out the client IP)
- Unit tests around AbstractLogin servlet - with necessary mocking to get a request to be successfully processed

I didn't refactor the 'parse out the client IP and compare to session IP' bit, as the servlet vs. the REST service implementation was very different - I didn't see a ton of value having a common library that all it does compare two strings - would still need the calling context to decide to where to pull the strings from.

Let me know your thoughts or what else you feel is needed for this PR.

File: azkaban-web-server/src/restli/java/azkaban/restli/ResourceUtils.java (hunk 1 above)
Comment (HappyRay) [inline azkaban-web-server/src/restli/java/azkaban/restli/ResourceUtils.java]: What about passing the client IP ( the last hop remote address ) into a common method? One call will be a value from req.getRemoteAddr() and another will be from context.getRawRequestContext().getLocalAttr("REMOTE_ADDR");

File: azkaban-web-server/src/test/java/azkaban/fixture/MockLoginAzkabanServlet.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
+package azkaban.fixture;
```
Comment (HappyRay) [inline azkaban-web-server/src/test/java/azkaban/fixture/MockLoginAzkabanServlet.java]: Pls add copyright notice.

File: azkaban-web-server/src/test/java/azkaban/webapp/servlet/LoginAbstractAzkabanServletTest.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        HttpServletResponse resp = getResponse(writer);
+
+
+        MockLoginAzkabanServlet servlet = MockLoginAzkabanServlet.getServletWithSession(sessionId,
+                "user", "127.0.0.1");
+
+
+        servlet.doPost(req, resp);
+
+        // Assert that our response was written (we have a valid session)
```
Comment (HappyRay) [inline azkaban-web-server/src/test/java/azkaban/webapp/servlet/LoginAbstractAzkabanServletTest.java]: Is the comment valid?

**Comment by mradamlacey** [comment]
Hi updated the pull request to refactor out a common method in `WebUtils` class to parse the real client ip.

Updated the spots in the Login servlet and the ResourceUtils class.

Updated unit tests.

Addressed your other comment about adding license file.

thanks!

File: azkaban-common/src/main/java/azkaban/utils/WebUtils.java [hunk 4]
```diff
@@ -164,4 +165,29 @@ else if (sizeBytes >= ONE_KB)
     else
       return sizeBytes + " B";
   }
+
+  public String getRealClientIpAddr(Map<String, String> httpHeaders, String remoteAddr){
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/utils/WebUtils.java]: Could you add JavaDoc?

File: azkaban-common/src/main/java/azkaban/utils/WebUtils.java [hunk 5]
```diff
@@ -164,4 +165,29 @@ else if (sizeBytes >= ONE_KB)
...
+
+    // If some upstream device added an X-Forwarded-For header
+    // use it for the client ip
+    // This will support scenarios where load balancers or gateways
+    // front the Azkaban web server and a changing Ip address invalidates
+    // the session
+
+    String clientIp = httpHeaders.getOrDefault("X-Forwarded-For", null);
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/utils/WebUtils.java]: Consider making "X-Forwarded-For" a constant?

File: azkaban-web-server/src/test/java/azkaban/fixture/MockLoginAzkabanServlet.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2014 LinkedIn Corp.
```
Comment (HappyRay) [inline azkaban-web-server/src/test/java/azkaban/fixture/MockLoginAzkabanServlet.java]: Make it 2016

**Comment by mradamlacey** [comment]
Just updated to address your latest changes.  Let me know if anything else is needed.

**Comment by HappyRay** [comment]
LGTM. Thanks!
One copyright notice date is still 2014. But I can merge it as it is.

**Pull request merged**